### PR TITLE
First draft of Logging feature implementation

### LINKED ISSUE #21: Logging

### PR DESCRIPTION
**Pull request by iskandersierra**
Fixes #21

## Description
Added module 'nact/monitoring' with configureLogging as the most important functionality. It is an extension for ActorSystem that includes a loggingEngine in the system and provides of a property log to the context of each message processing.

## Related Issue
This PR gives solution to issue #21.

## Motivation and Context
I would like to participate in the development of this project and took a simple 'up for grabs' issue to start familiarizing with the code.

## How Has This Been Tested?
This feature was tested with unit testing using mocha, chai and sinon.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: lib/monitoring/logging-actor.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  if (!engine || engine === logNothing) {
+    return { ...ctx, log: logNothing };
+  }
+
+  const subEngine = new MessageContextLoggingEngine(engine, actor.path);
+  return { ...ctx, log: subEngine };
+};
+
+module.exports = {
+  // spawnLoggingActor,
```
Comment (ncthbrt) [inline lib/monitoring/logging-actor.js]: Mind removing commented out code please

File: lib/monitoring/logging-engine.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+const logLevelAsText = [
+  'OFF',
+  'TRACE',
+  'DEBUG',
+  'INFO',
+  'WARNING',
+  'ERROR',
+  'CRITICAL'
+];
```
Comment (ncthbrt) [inline lib/monitoring/logging-engine.js]: I've been using `Symbol` for this purpose. E.g. 
```const SupervisionActions = {
  stop: Symbol('stop'),
  stopAll: Symbol('stopAll'),
  escalate: Symbol('escalate'),
  resume: Symbol('resume'),
  reset: Symbol('reset'),
  resetAll: Symbol('resetAll')
};```

File: lib/monitoring/logging-engine.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+      (logEvent =>
+        ''.concat(
+          '[',
+          logLevelToString(logEvent.level),
+          '] ',
+          logEvent.category,
+          ': ',
+          logEvent.message
+        ));
+      // `[{logLevelToString(level)}] {logEvent.category}: {logEvent.message}`
```
Comment (ncthbrt) [inline lib/monitoring/logging-engine.js]: More commented out code

File: lib/monitoring/logging-engine.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  }
+
+  log (logEvent) {
+    const extra = [logEvent.properties, logEvent.metrics].filter(x => !!x);
+    const channel =
+      this.channels[logEvent.level] ||
+      (logEvent.level >= LogLevel.ERROR
```
Comment (ncthbrt) [inline lib/monitoring/logging-engine.js]: Ah I see why you're using numbers

File: lib/monitoring/logging-engine.js (hunk 2 above)
Comment (ncthbrt) [reply lib/monitoring/logging-engine.js]: But I see that you're using the values in the LogLevel to define a natural ordering

File: lib/monitoring/logging-actor.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
+const {
```
Comment (ncthbrt) [inline lib/monitoring/logging-actor.js]: Strictly speaking, this filename is misleading, it isn't a "logging actor".

File: lib/monitoring/logging-engine.js [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+  'TRACE',
+  'DEBUG',
+  'INFO',
+  'WARNING',
+  'ERROR',
+  'CRITICAL'
+];
+
+function logLevelToString (level) {
+  return logLevelAsText[level] || '???';
```
Comment (ncthbrt) [inline lib/monitoring/logging-engine.js]: Do you think '???' is the correct text to return here?

File: lib/monitoring/logging-engine.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+class NoopLoggingEngine extends AbstractLoggingEngine {
+  log (logEvent) {}
+}
+
+const logNothing = new NoopLoggingEngine();
+
+class ConsoleLoggingEngine extends AbstractLoggingEngine {
+  constructor (consoleProxy, formatter) {
+    super();
+    this.channels = [
```
Comment (ncthbrt) [inline lib/monitoring/logging-engine.js]: This is cool

File: lib/monitoring/logging-actor.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+      logEvent.category,
+      this.actorPath.toString().concat(': ', logEvent.message),
+      logEvent.properties,
+      logEvent.metrics
+    );
+    this.engine.log(logEvent);
+  }
+}
+
+const prepareContextForLogging = (actor, ctx) => {
```
Comment (ncthbrt) [inline lib/monitoring/logging-actor.js]: I'd be happier if you moved this maybe into the base `actor.js` file with the rest of the context stuff.

File: lib/monitoring/logging-engine.js (hunk 2 above)
Comment (ncthbrt) [reply lib/monitoring/logging-engine.js]: So ignore this

File: lib/monitoring/logging-actor.js (hunk 5 above)
Comment (iskandersierra) [reply lib/monitoring/logging-actor.js]: You are completelly right here. The issue is that I started implementing the logging-actor to do this job.  But some weird circularity dependency problems was provoking that spawnStateless were still undefined wher I wanted to use it. Si I went for the simplest implementation for the moment until you give feedback. Some commented code you see is about this previous failed implementation. If you have some thouhgs about it I can update this with a Actor-based logging implementation.

File: lib/monitoring/logging-actor.js (hunk 8 above)
Comment (iskandersierra) [reply lib/monitoring/logging-actor.js]: Ok, no problem.

File: lib/monitoring/logging-actor.js (hunk 1 above)
Comment (iskandersierra) [reply lib/monitoring/logging-actor.js]: I will wait for comments to decide whether this file stays or not.

File: lib/monitoring/logging-engine.js (hunk 6 above)
Comment (iskandersierra) [reply lib/monitoring/logging-engine.js]: Not at all but got out of ideas :D Do you have any suggestion?

File: lib/monitoring/logging-engine.js (hunk 3 above)
Comment (iskandersierra) [reply lib/monitoring/logging-engine.js]: I thought string interpolation was a thing, but didnt work. For me, used to TS, it is some times difficult to know the features available in JS

**Comment by iskandersierra** [comment]
About your comments to the PR:

> Do you thinking log-level filtering should be the responsibility of each engine implementation?

I totally agree. The problem is that there are dozens of really great logging libraries out there. I think a minimalistic approach is much better here, accompanied with our own contrib-like projects to integrate nact with other technologies. And the cleaner abstraction we give in nact for logging (and other transversal stuff like logging, configuration, transport, etc.) the easies it is to integrate with other libraries.

> Sometimes having multiple logging engines is useful. E.g. if critical error make a big noise, wake everyone up, if warning, post to slack, etc.

Similar to the previous comment, I think this should not be responsibility of nact. It would be a bulk of code outside the scope of nact. We can select a couple of good and well supported logging libraries and start the corresponding nact-logging-<<provider>>-contrib projects (leave naming to you :D) as a good suggestion. But I believe the simplest console implementation should be the default.

> When volume of logging is high, how would this implementation make it easy to handle congestion control and batching? An alternative approach which shouldn't be a huge change, would be to define the logging engine as a stateful actor (with the system as the parent). This would make it easier to do batching and is nice dog-fooding.

In this I'm divided. It is true that an actor could implement this decisions. But again I think that this should be a responsibility of the chosen library. If it support batching, dropping, filtering, broadcasting, good for you. But to load nact with that much code is a risk in my opinion. 

I still can put an actor in the receiving part of the logs only to redirect the call to the real engine implementation, and I see that as a nice upgrade. But nothing else in mi opinion should be implemented in nact, to reduce the risk of baggage code.

> Also just wanted to say that you've got some really great tests.

Thanks, I try my best!

File: lib/monitoring/logging-engine.js (hunk 6 above)
Comment (ncthbrt) [reply lib/monitoring/logging-engine.js]: Could return simply undefined?

File: lib/monitoring/logging-engine.js (hunk 3 above)
Comment (ncthbrt) [reply lib/monitoring/logging-engine.js]: String interpolation is available. It looks like this:
````
`Hello world, my name is ${name}`
```

**Comment by ncthbrt** [comment]
> I still can put an actor in the receiving part of the logs only to redirect the call to the real engine implementation, and I see that as a nice upgrade. But nothing else in mi opinion should be implemented in nact, to reduce the risk of baggage code.

If I understand you correctly, as an example, the console logging actor would look somthing like the following:

```js
const channels = 
    this.channels = [
      null,
      consoleProxy.trace,
      consoleProxy.debug,
      consoleProxy.info,
      consoleProxy.warn,
      consoleProxy.error
    ];

const spawnConsoleLogger = system => 
    spawn(
        system,
        (state={}, logEvent, ctx) => {
            const msg =`[${logLevelToString(logEvent.level)}] ${logEvent.category}: ${logEvent.message}`;
            const extra = [logEvent.properties, logEvent.metrics].filter(x => !!x);
            const channel = this.channels[logEvent.level] || ()=>{};
            channel(msg, ...extra);     
        },
        "logger"
    );

**Comment by iskandersierra** [comment]
It would be like this:
``` javascript
class ActorLoggingEngine extends AbstractLoggingEngine {
  constructor (loggingActorRef) {
    super();
    this.loggingActorRef = loggingActorRef;
    const { dispatch } = require('../functions');
    this.dispatch = dispatch;
  }

  log (logEvent) {
    this.dispatch(this.loggingActorRef, logEvent);
  }
}

const spawnLoggingActor = (system, engine) => {
  if (!(engine instanceof AbstractLoggingEngine)) {
    throw new Error('Expected a LoginEngine.');
  }

  const { spawnStateless } = require('../actor');

  return spawnStateless(
    system,
    (msg, ctx) => {
      if (msg instanceof LogEvent) {
        engine.log(msg);
      }
      // else ??? Is there a Unhandled/DeadLetters channel?
    },
    'logger'
  );
};
```
I had to put the ```const { dispatch } = require('../functions');``` statement inside the function, otherwise dispatch was getting undefined (as well as spawnStateless).

This actor can use any engine implementation, including ConsoleLoggingEngine. The spawnLoggingActor is for internal use in configure logging, as such:
```javascript
const configureLogging = (engine) => (system) => {
  if (system.loggingEngine) {
    throw new Error('Do not configure logging more than once. The behavior could be undefined otherwise.');
  }

  if (!(engine instanceof AbstractLoggingEngine)) {
    throw new Error('Must use a valid logging engine. Try with logToConsole() from module "nact/monitoring".');
  }

  console.log('********************** system reference type is: ', system.reference.constructor.name);
  try {
    const loggingActor = spawnLoggingActor(system.reference, engine);
    const loggingEngine = new ActorLoggingEngine(loggingActor);

    console.log('********************** Got logging engine: ', loggingEngine.constructor.name);
    return Object.assign(system, { loggingEngine });
  } catch (error) {
    console.log('********************** Error getting logging engine: ', error);
  }
};
```

However I'm getting an error here that I hope you can help me to find.

> ********************** system reference type is:  ActorSystemReference
********************** Error getting logging engine:  **Error: Actor has stopped or never even existed**
    at Object.module.exports.applyOrThrowIfStopped (C:\Dev\git\nact\lib\system-map.js:1:3589)
    at spawn (C:\Dev\git\nact\lib\actor.js:2:3509)
    at spawnStateless (C:\Dev\git\nact\lib\actor.js:2:3856)
    at spawnLoggingActor (C:\Dev\git\nact\lib\monitoring\logging-actor.js:1:3556)
    at system (C:\Dev\git\nact\lib\monitoring\index.js:1:3319)
    at ActorSystem.forEach.extension (C:\Dev\git\nact\lib\system.js:1:10075)
    at Array.forEach (<anonymous>)
    at new ActorSystem (C:\Dev\git\nact\lib\system.js:1:10005)
    at start (C:\Dev\git\nact\lib\system.js:1:12394)
    at Context.it (C:\Dev\git\nact\test\logging-actor.js:50:14)
    at callFn (C:\Dev\git\nact\node_modules\mocha\lib\runnable.js:354:21)
    at Test.Runnable.run (C:\Dev\git\nact\node_modules\mocha\lib\runnable.js:346:7)
    at Runner.runTest (C:\Dev\git\nact\node_modules\mocha\lib\runner.js:442:10)
    at C:\Dev\git\nact\node_modules\mocha\lib\runner.js:560:12
    at next (C:\Dev\git\nact\node_modules\mocha\lib\runner.js:356:14)
    at C:\Dev\git\nact\node_modules\mocha\lib\runner.js:366:7
    at next (C:\Dev\git\nact\node_modules\mocha\lib\runner.js:290:14)
    at Immediate.<anonymous> (C:\Dev\git\nact\node_modules\mocha\lib\runner.js:334:5)
    at runCallback (timers.js:789:20)
    at tryOnImmediate (timers.js:751:5)
    at processImmediate [as _immediateCallback] (timers.js:722:5)

File: lib/monitoring/index.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  AbstractLoggingEngine,
+  logToConsole
+} = require('./logging-engine');
+
+const configureLogging = (engine) => (system) => {
+  if (system.loggingEngine) {
+    throw new Error('Do not configure logging more than once. The behavior could be undefined otherwise.');
+  }
+
+  if (!(engine instanceof AbstractLoggingEngine)) {
```
Comment (ncthbrt) [inline lib/monitoring/index.js]: @iskandersierra I initially was doing type checks, but it turns out it causes problems with plugins such as the postgres persistence engine: It forces all dependent libraries to upgrade their nact versions in lock step.

Comment (iskandersierra) [reply lib/monitoring/index.js]: Fair. I changed it by:

``` javascript
  if (!engine) {
    throw new Error('Expected a LoginEngine.');
  }
```

Any comments on the previous issue on configureLogging?

Comment (ncthbrt) [reply lib/monitoring/index.js]: @iskandersierra I've merged an update which should fix the error below:
``********************** system reference type is: ActorSystemReference
********************** Error getting logging engine: Error: Actor has stopped or never even existed```

The issue was that the system was being added to the system map AFTER the extensions were being configured.

**Comment by ncthbrt** [comment]
Using `spawnLoggingActor` internally isn't quite what I was getting at, think it creates additional overhead and abstraction. 

What I meant was removing the AbstractLoggingEngine stuff and replacing the engine abstraction with  an actor. `ctx.log.error("Something bad happened")` for example would then simply dispatch to this actor. You could easily write adapters for the most popular frameworks using the actor. 

I know that's what I've done with the persistence engine, but that is because it breaks abstraction boundaries (I kind of still would like to rethink that design).

That would change configure logging to (where engine is a function which takes in the system and returns an actor reference)

```js
const configureLogging = (engine) => (system) => {
  if (system.loggingEngine) {
    throw new Error('Do not configure logging more than once. The behavior could be undefined otherwise.');
  }
  try {
    const loggingActor = engine(system.reference);
    system.loggingEngine = loggingActor;
  } catch (error) {
    console.log('********************** Error getting logging engine: ', error);
  }
};
```

File: lib/monitoring/index.js (hunk 9 above)
Comment (iskandersierra) [reply lib/monitoring/index.js]: Great! Now it is failing further away, in this code:
> TypeError: Cannot read property 'system' of undefined
    at Object.module.exports.applyOrThrowIfStopped (C:\Dev\git\nact\lib\system-map.js:1:3402)
    at spawn (C:\Dev\git\nact\lib\actor.js:2:3509)
    at spawnStateless (C:\Dev\git\nact\lib\actor.js:2:3856)
    at spawnLoggingActor (C:\Dev\git\nact\lib\monitoring\logging-actor.js:1:3940)
    at system (C:\Dev\git\nact\lib\monitoring\index.js:1:3098)
    at C:\Dev\git\nact\test\logging-actor.js:39:40
``` javascript
const spawn = (parent, f, name, properties) =>
  systemMap.applyOrThrowIfStopped(parent, p => p.assertNotStopped() && new Actor(p, name, p.system, f, properties).reference);
```

It seems that in ```p.system```, p is giving me undefined.

**Comment by iskandersierra** [comment]
Agree. I'm removing the unnecesary abstractions.

File: lib/monitoring/index.js (hunk 9 above)
Comment (ncthbrt) [reply lib/monitoring/index.js]: Surely p.assertNotStopped() would also fail in that case?

**Comment by ncthbrt** [comment]
@iskandersierra Thank you. Apologise if I'm being full of nonsense :)

**Comment by ncthbrt** [comment]
@iskandersierra Just to check, were you able to solve the `TypeError: Cannot read property 'system' of undefined` problem?

**Comment by iskandersierra** [comment]
Hello @ncthbrt 
I made the changes to simplify the abstractions. It took me a while because I'm interleaving with my job.
Everything is working as expected, no more ```TypeError: Cannot read property 'system' of undefined``` problem.

File: lib/monitoring/console-engine.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+  channels[LogLevel.WARN] = consoleProxy.warn || channels[LogLevel.INFO];
+  channels[LogLevel.ERROR] = consoleProxy.error || channels[LogLevel.WARN];
+  channels[LogLevel.CRITICAL] = consoleProxy.critical || channels[LogLevel.ERROR];
+
+  const format = formatter || ((logEvent) =>
+    `[${logLevelToString(logEvent.level)}] ${logEvent.category}: ${logEvent.message}`);
+
+  return (system) => spawnStateless(
+    system,
+    (logEvent, ctx) => {
```
Comment (ncthbrt) [inline lib/monitoring/console-engine.js]: 👍

File: test/logging-engine.js [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+
+describe('LoggingFacade', () => {
+  it('should not be able to be instantiated', () => {
+    (() => new LoggingFacade()).should.throw(Error);
+  });
+});
+
+describe('LoggingFacadeImpl', () => {
+  it('should be able to be instantiated', () => {
+    (new LoggingFacadeImpl(() => {})).should.been.instanceOf(LoggingFacadeImpl);
```
Comment (ncthbrt) [inline test/logging-engine.js]: been? is that part of chai?

**Comment by ncthbrt** [comment]
Looks good. I'm going to merge this if you're happy @iskandersierra 
Btw are you on the nact discord? Would like to ping you when your changes are released (need to get them into the reason version too)

File: test/logging-engine.js (hunk 11 above)
Comment (iskandersierra) [reply test/logging-engine.js]: It is a mistake but sinon for chai introduces the chain been as same to be so it didn't failed.

**Comment by iskandersierra** [comment]
@ncthbrt Feel free to merge. Next step could be integrating with a couple of most used logging providers. If you have some preferences I would like to hear them.

**Comment by ncthbrt** [comment]
Think we're using Winston at work?

**Pull request merged**